File: python/wampprotofbs/parsers/abort.py
from typing import Any

import flatbuffers
from wampproto.messages.abort import Abort, IAbortFields
from wampproto.serializers.payload import serialize_payload

from wampprotofbs.parsers import helpers
from wampprotofbs.gen import Abort as Abort_FBS


class AbortFields(IAbortFields):
    def __init__(self, gen: Abort_FBS.Abort, payload: bytes):
        self._gen = gen
        self._ex = helpers.PayloadExpander(payload, gen.PayloadSerializerId())

    @property
    def reason(self) -> str:
        return self._gen.Reason().decode("utf-8")

    @property
    def details(self) -> dict[str, Any]:
        return {}

    @property
    def args(self) -> list[Any]:
        return self._ex.args

    @property
    def kwargs(self) -> dict[str, Any]:
        return self._ex.kwargs

    @property
    def payload_serializer(self) -> int:
        return self._gen.PayloadSerializerId()

    @property
    def payload(self) -> bytes:
        return self._ex.payload


def abort_to_flatbuffer(a: Abort) -> bytes:
    builder = flatbuffers.Builder(1024)

    reason_offset = builder.CreateString(a.reason)
    payload_serializer = helpers.select_payload_serializer(a.details)

    Abort_FBS.AbortStart(builder)
    Abort_FBS.AbortAddReason(builder, reason_offset)
    Abort_FBS.AbortAddPayloadSerializerId(builder, payload_serializer)
    abort_msg = Abort_FBS.AbortEnd(builder)

    builder.Finish(abort_msg)
    payload = serialize_payload(payload_serializer, a.args, a.kwargs)

    return helpers.prepend_header(Abort.TYPE, bytes(builder.Output()), payload)


def flatbuffer_to_abort(data: bytes) -> Abort:
    message_data, payload_data = helpers.extract_message(data)
    fb_abort = Abort_FBS.Abort.GetRootAsAbort(message_data, 0)

    return Abort(AbortFields(fb_abort, payload_data))

File: go/parsers/abort.go
package parsers

import (
	"github.com/google/flatbuffers/go"

	"github.com/xconnio/wampproto-flatbuffers/go/gen"
	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-go/serializers"
)

type Abort struct {
	gen *gen.Abort
	ex  *PayloadExpander
}

func NewAbortFields(g *gen.Abort, payload []byte) messages.AbortFields {
	return &Abort{
		gen: g,
		ex:  &PayloadExpander{payload: payload, serializer: g.PayloadSerializerId()},
	}
}

func (a *Abort) Reason() string {
	return string(a.gen.Reason())
}

func (a *Abort) Details() map[string]any {
	return map[string]any{}
}

func (a *Abort) Args() []any {
	return a.ex.Args()
}

func (a *Abort) KwArgs() map[string]any {
	return a.ex.Kwargs()
}

func AbortToFlatbuffers(m *messages.Abort) ([]byte, error) {
	builder := flatbuffers.NewBuilder(256)

	reason := builder.CreateString(m.Reason())

	gen.AbortStart(builder)
	gen.AbortAddReason(builder, reason)
	payloadSerializer := selectPayloadSerializer(m.Details())
	gen.AbortAddPayloadSerializerId(builder, payloadSerializer)
	abort := gen.AbortEnd(builder)

	builder.Finish(abort)

	payload, err := serializers.SerializePayload(payloadSerializer, m.Args(), m.KwArgs())
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeAbort, builder.FinishedBytes(), payload), nil
}

func FlatbuffersToAbort(data, payload []byte) (*messages.Abort, error) {
	abort := gen.GetRootAsAbort(data, 0)
	return messages.NewAbortWithFields(NewAbortFields(abort, payload)), nil
}
